### PR TITLE
Improve service documentation

### PR DESCRIPTION
**Pull request by hlbarber**
## Motivation and Context

A few areas of documentation could be improved.

## Description

- Add module documentation to `plugin.rs`.
- Remove reference to `FromRequest` in `FromParts`.
- Improve `Route` documentation.

### CODE REVIEW

File: rust-runtime/aws-smithy-http-server/src/plugin/mod.rs [hunk 1]
```diff
@@ -3,6 +3,11 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+//! The plugin system allows for parameterized modification of [`Operation`]s during the build procedure.
```
Comment (LukeMathWalker) [inline rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:6]: This feels a bit abstract 🤔

Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:6]: Yeah, any thoughts?

Perhaps a doc test with some annotations will make it more concrete?

Comment (LukeMathWalker) [reply rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:6]: A doc test would certainly help.

Perhaps

```
The plugin system allows you to build middlewares that have access to the details of the [`Operation`] that was invoked (e.g. its name).
```

Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/src/plugin/mod.rs]: Hows this look?

https://github.com/awslabs/smithy-rs/pull/2020/commits/d717444a0f0fd5e9c82edca190a217c2168e6d2b

I'm wondering whether it'd be nice to demonstrate the extension trait? Should I pull in `InstrumentExt` in an example?

File: rust-runtime/aws-smithy-http-server/src/plugin/mod.rs [hunk 2]
```diff
@@ -3,10 +3,47 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-//! The plugin system allows for parameterized modification of [`Operation`]s during the build procedure.
+//! The plugin system allows the user to build middleware with an awareness of the operation its applied to.
```
Comment (LukeMathWalker) [inline rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:6]:
```suggestion
//! The plugin system allows the user to build middleware with an awareness of the operation it is applied to.
```

File: rust-runtime/aws-smithy-http-server/src/plugin/mod.rs [hunk 3]
```diff
@@ -3,10 +3,47 @@
...
+//! let plugin = plugin_from_operation_name_fn(|name| PrintLayer { name });
+//! ```
+//!
+//! # Combine [`Plugin`]s
+//!
+//! ```
+//! # use aws_smithy_http_server::plugin::*;
+//! # let a = (); let b = ();
+//! // Combine `Plugin`s `a` and `b`
+//! let plugin = PluginPipeline::empty()
```
Comment (LukeMathWalker) [inline rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:42]:
```suggestion
//! let plugin = PluginPipeline::new()
```

File: rust-runtime/aws-smithy-http-server/src/plugin/mod.rs (hunk 1 above)
Comment (LukeMathWalker) [reply rust-runtime/aws-smithy-http-server/src/plugin/mod.rs]: That looks good!
I would avoid duplicating too much content between the module docs and the `PluginPipeline` docs.

File: rust-runtime/aws-smithy-http-server/src/plugin/mod.rs [hunk 4]
```diff
@@ -3,6 +3,48 @@
...
+//! let plugin = plugin_from_operation_name_fn(|name| PrintLayer { name });
+//! ```
+//!
+//! # Combine [`Plugin`]s
+//!
+//! ```
+//! # use aws_smithy_http_server::plugin::*;
+//! # let a = (); let b = ();
+//! // Combine `Plugin`s `a` and `b`
+//! let plugin = PluginPipeline::new()
```
Comment (82marbag) [inline rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:42]: What's the order in which they will run?

Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:42]: Good point, I'll document that

Comment (hlbarber) [reply rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:42]: 9cdb8209f0636fec73ba8867df2b6dd9cc79d439

File: rust-runtime/aws-smithy-http-server/src/plugin/mod.rs [hunk 5]
```diff
@@ -3,6 +3,51 @@
...
+//! # Filtered application of a HTTP [`Layer`](tower::Layer)
+//!
+//! ```
+//! # use aws_smithy_http_server::plugin::*;
+//! # let layer = ();
+//! // Create a `Plugin` from a HTTP `Layer`
+//! let plugin = HttpLayer(layer);
+//!
+//! // Only apply the layer to operations with name "GetPokemonSpecies"
+//! let plugin = filter_by_operation_name(plugin, |name| name == "GetPokemonSpecies");
```
Comment (LukeMathWalker) [inline rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:20]: Let's use the `::NAME` property on an operation ZST here.

File: rust-runtime/aws-smithy-http-server/src/plugin/mod.rs [hunk 6]
```diff
@@ -3,6 +3,51 @@
...
+//! ```
+//!
+//! # Construct [`Plugin`] from Operation name closure
```
Comment (LukeMathWalker) [inline rust-runtime/aws-smithy-http-server/src/plugin/mod.rs:23]:
```suggestion
//! # Construct a [`Plugin`] from a closure that takes as input the operation name
```

File: rust-runtime/aws-smithy-http-server/src/routing/route.rs [hunk 7]
```diff
@@ -46,12 +46,15 @@ use tower::{
     Service, ServiceExt,
 };
 
-/// How routes are stored inside a [`Router`](super::Router).
+/// A HTTP [`Service`] representing a single route.
+///
+/// The construction of [`Route`] from a named HTTP [`Service`] S, erases the type of `S`.
```
Comment (LukeMathWalker) [inline rust-runtime/aws-smithy-http-server/src/routing/route.rs:51]:
```suggestion
/// The construction of [`Route`] from a named HTTP [`Service`] `S`, erases the type of `S`.
```

**Review by LukeMathWalker — APPROVED** [review]

**Pull request merged**